_posts/2019-07-30-indexing-searching-nuget-with-azure-functions-and-search.markdown

```diff
@@ -0,0 +1,891 @@
+
+Success! Our indexer seems to work, and search returns results!
+
+### One more thing...
+
+Ideally, our search index would score packages by download count to help determine relevance. Unfortunately, this one important detail is currently missing from the NuGet V3 API... We've made a [feature request to expose download counts](https://github.com/NuGet/NuGetGallery/issues/3532), but right now download counts are not available (yet?)...
+
+Now if that data ever is made available, how would we add it to the index? Re-index every package every time we get new download counts? That would be overkill. Azure Search can "merge" data, so based on the identifier we created in `PackageDocument`, we could merge in download counts without having to touch the other fields for a package that has already been indexed.
+
+## Conclusion
+
+We have built a pipeline of functions to index packages from NuGet.org, and their public types. We added an API to search that index, and made it compatible with existing ReSharper and Rider versions.
+Source code for this blog post [is on GitHub](https://github.com/maartenba/NuGetTypeSearch).
+
+We tried a deployment of these functions and did a full index of all of NuGet.org's public types and namespaces. A full index run back in May 2019 took ~12h to complete when scaled on 2x [B1ms reserved instances](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/sizes-general#b-series) - that's one vCPU and 2 GB of RAM. Those 12h could be sped up on larger instances, or on more instances where indexing/downloading packages can run in parallel.
+
+We found 2.1 million package versions in the NuGet.org catalog, over 8400 catalog pages with 4.2 million catalog leafs. Double? Remember repository signing happened, so roughly double makes sense such a short time after that happened...
+
+Will this go in production? No idea. It was definitely fun to build, I've also taken some learnings from working with Azure Functions into [Rider's Azure plugin](https://plugins.jetbrains.com/plugin/11220-azure-toolkit-for-rider) - such as the ability to quickly run/debug individual functions while developing which is super handy! (will be in the 2019.2 release)
+
+![Run individual Azure Function with Rider](../images/2019/07/run-individual-function.png)
+
+Regarding the indexing/search pipeline, I would deploy each task into a separate Azure Function App, for isolation and scaling purposes. The first function (catalog traversal) is the only one that has to be always on, because of the custom trigger. It doesn't need a lot of capacity, so it could easily be run on the smallest VM that can be used. Indexing could be a separate, consumption-based deployment, that would scale out and in based on queue size. And then the search API endpoint would also be separate, for scale, but also for isolation from potential issues in the indexing process.
```

Comment:
```suggestion
**Regarding the indexing/search pipeline,** I would deploy each task into a separate Azure Function App, for isolation and scaling purposes. The first function (catalog traversal) is the only one that has to be always on, because of the custom trigger. It doesn't need a lot of capacity, so it could easily be run on the smallest VM that can be used. Indexing could be a separate, consumption-based deployment, that would scale out and in based on queue size. And then the search API endpoint would also be separate, for scale, but also for isolation from potential issues in the indexing process.
```